tensorboard/defs/py_repl.bzl

```diff
@@ -0,0 +1,105 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Defines `py_repl` rule for creating a `bazel run`-able python REPL."""
+
+# Minimal wrapper over ctx.actions.write to write a string to a file.
+# Simpler than the genrule equivalent since we don't need to escape.
+def _write_file_impl(ctx):
+    ctx.actions.write(ctx.outputs.out, ctx.attr.content)
+
+_write_file = rule(
+    implementation = _write_file_impl,
+    attrs = {
+        "out": attr.output(mandatory = True),
+        "content": attr.string(),
+    },
+)
+
+def _dedent(text):
+    """Like `textwrap.dedent()` but only supports space indents."""
+    indents = []
+    for line in text.splitlines():
+        stripped = line.lstrip(" ")
+        if not stripped:
+            continue
+        indents.append(len(line) - len(stripped))
+    if not indents:
+        return
```

Comment:
`return ""`, not just `return`?

Reply:
Thanks - `return text` actually is what I intended when input is all whitespace.  (In this context the `rstrip()` on the way in makes it equivalent, but in the general case I think probably better to preserve.)